notebooks/logging-examples/logging-experiment-failures.ipynb

```diff
@@ -0,0 +1,157 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "3b466d92",
+   "metadata": {},
+   "source": [
+    "## Distinguishing Failed Experiments with Rubicon \n",
+    "Modeling runs can fail for a number of reasons. When logging with rubicon_ml, a failed run may result in an empty or incomplete experiment. In this example, we'll walk through how to handle such experiments."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d4f58990",
+   "metadata": {},
+   "source": [
+    "First lets simulate the problem. To do this we'll create an estimator that will fail on it's `fit()` 30% of the time.  We will consider any experiment that has a `none` value for `self.state`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "id": "f9a506c3",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sklearn.base import BaseEstimator\n",
+    "import random\n",
+    "\n",
+    "class BadEstimator(BaseEstimator):\n",
+    "    def __init__(self, state=None):\n",
+    "        super().__init__()\n",
+    "        self.state=None\n",
+    "        self.knn = KNeighborsClassifier(n_neighbors=2)\n",
+    "    def fit(self, X, y):\n",
+    "        self.knn.fit(X, y)\n",
+    "        output=random.random()\n",
+    "        if output<.3:\n",
+    "            output=None\n",
+    "        self.state=output\n",
+    "    def score(self, X):\n",
+    "        knn_score = self.knn.score(X)\n",
+    "        return knn_score"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2d62eaa1",
+   "metadata": {},
+   "source": [
+    "Now let's attempt to create a `rubicon_ml.sklearn` pipeline with this failing estimator and attempt to `fit` the pipeline twenty times. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "id": "7ccb023c",
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "from rubicon_ml.sklearn import make_pipeline\n",
+    "from sklearn.neighbors import KNeighborsClassifier\n",
+    "from sklearn.impute import SimpleImputer\n",
+    "from rubicon_ml import Rubicon\n",
+    "\n",
+    "random.seed(18)\n",
+    "\n",
+    "rubicon = Rubicon(\n",
+    "    persistence=\"memory\",\n",
+    ")\n",
+    "project = rubicon.get_or_create_project(name=\"Failed Experiments\")\n",
+    "\n",
+    "X = [[1], [1], [1], [1]]\n",
+    "y = [1, 1, 1, 1]\n",
+    "pipe=make_pipeline(project, SimpleImputer(strategy=\"mean\"),BadEstimator())\n",
+    "for _ in range(20):\n",
```

Comment:
can we move this for loop into the next cell and combine it with that loop? something like this in the following cell:

```
for _ in range(20):
    pipe.fit(X, y)

    if not hasattr(pipe, "state_"):
        pipe.experiment.add_tags(["failed"])

Reply:
Isn't state an attribute of the estimator itself, not the pipeline? Also, when I don't have it in the constructor I don't see it logged

Reply:
yeah my bad, you'll need to access it from the estimator. so something like `pipeline.stages[0].state_` or `pipeline["stage_name"].state_`